# Zika_Project

This repository contains all code and utilities used in the publication: "Hunting unannotated human proteins to better understand Zika virus infection" by Leblanc S & Brunet MA.
The manuscript presents a computational framework for identification of novel proteins and is based on the re-analysis of transcriptomics and proteomics datasets.

![alt text](https://github.com/MAB-Lab/Zika_Project/Graphical_abstract.jpg?raw=true)

# Abstract:
The Zika virus is a flavivirus that can cause fulminant outbreaks and lead to Guillain-Barr√© syndrome, microcephaly and fetal demise. Like other flaviviruses, the Zika virus is transmitted by mosquitoes and provokes neurological disorders. Despite its risk to public health, no antiviral nor vaccine are currently available. In the recent years, several studies have set to identify human host proteins interacting with Zika viral proteins to better understand its pathogenicity. Yet these studies used standard human protein sequence databases. Such databases rely on genome annotations, which enforce a minimal ORF length criterion. An ever-increasing number of studies have demonstrated the shortcomings of such annotation with thousands of functional ORFs lacking. Here we show that the use of a customized database including currently non-annotated proteins led to the identification of 3 alternative proteins as interactors of the viral capsid and NS4A proteins. Furthermore, 5 alternative proteins were identified in the proteome profiling of Zika infected monocytes, one of which was significantly up-regulated. This study presents a computational framework for the re-analysis of proteomics dataset to better investigate the viral-host protein interplays upon infection with the Zika virus.

# Contents of repository:
